syne_tune/remote/remote_main.py

```diff
@@ -27,11 +28,14 @@
     parser.add_argument('--store_logs', dest='store_logs', action='store_true', default=False)
     parser.add_argument('--log_level', type=int, default=logging.INFO)
     parser.add_argument('--no_tuner_logging', type=str, default='False')
+    parser.add_argument('--region_name', type=str, default='us-west-2')
```

Comment:
It seems to me that we can avoid adding this extra argument and case in `remote_main.py` by just setting the environment variable in the caller script.

The following would be equivalent:
``` 

        # avoids error "Must setup local AWS configuration with a region supported by SageMaker."
        # in case no region is explicitely configured by providing a default region
        environment = estimator_kwargs.get("environment", {})
        if "AWS_DEFAULT_REGION" not in environment:
          environment["AWS_DEFAULT_REGION"] = boto3.Session().region_name
        tuner_estimator = PyTorch(
            # path which calls the tuner
            entry_point="remote_main.py",
            source_dir=str(self.remote_script_dir()),
            instance_type=self.instance_type,
            instance_count=1,
            role=self.role,
            py_version="py3",
            framework_version='1.6',
            image_uri=self.syne_tune_image_uri(),
            hyperparameters=hyperparameters,
            checkpoint_s3_uri=checkpoint_s3_uri,
            environment=environment,
            **self.estimator_kwargs,
        )
```

but does not require any change in`remote_main.py` while requiring a similar change in `remote_tuner.py` (e.g. setting an env variable instead of a hyperparameter).

Could we make this change unless you see a downside of this approach?

Reply:
I experimented with this idea and I didn't manage to get it to work. My understanding is that `remote_main.py` loads a `Tuner` instance, that triggers deserialization of all underlying objects: https://github.com/awslabs/syne-tune/blob/main/syne_tune/remote/remote_main.py#L37

`SagemakerBackend` on loading creates a new `sagemaker.Session()` instance: https://github.com/awslabs/syne-tune/blob/main/syne_tune/backend/sagemaker_backend/sagemaker_backend.py#L230

That overrides the region that was set before: https://github.com/awslabs/syne-tune/blob/main/syne_tune/backend/sagemaker_backend/sagemaker_backend.py#L251

Since this code is getting executed on a remote Sagemaker instance, where neither `AWS_DEFAULT_REGION` nor `awscli` config are available, it falls back to the default here: https://github.com/awslabs/syne-tune/blob/main/syne_tune/backend/sagemaker_backend/sagemaker_backend.py#L244

The only workaround I could come up with is propagating `region_name` with a CLI argument. I think another potential solution is to query a special AWS EC2 service as suggested here: https://stackoverflow.com/a/9735663 However, I don't know if this URL is available for Sagemaker containers.

Reply:
Thanks for the detailed answer and the investigation. Let's ship the version as it is as it is functional and perhaps we can try to do this change later.

Regarding your analysis, it is not 100% clear to me why the solution with the environment variable would not work. In particular the following point:

> Since this code is getting executed on a remote Sagemaker instance where neither AWS_DEFAULT_REGION nor awscli config are available

is not clear to me because setting `environment={"foo": "A3}` should set the environment variable "foo=A3" on the remote instance.

Reply:
For instance, I just tried this standalone script which launches an estimator after setting the `AWS_DEFAULT_REGION` environment variable. 
When running on my machine the script displays "us-west-2" (default region on my machine) and "us-east-1" (region on the remote machine set through the environment variable).

```python
import os
from pathlib import Path

import boto3
from sagemaker.pytorch import PyTorch

def get_execution_role():
    """
    :return: sagemaker execution role that is specified with the environment variable `AWS_ROLE`, if not specified then
    we infer it by searching for the role associated to Sagemaker. Note that
    `import sagemaker; sagemaker.get_execution_role()`
    does not return the right role outside of a Sagemaker notebook.
    """
    client = boto3.client("iam")
    sm_roles = client.list_roles(PathPrefix="/service-role/")['Roles']
    for role in sm_roles:
        if 'AmazonSageMaker-ExecutionRole' in role['RoleName']:
            return role['Arn']

if __name__ == '__main__':
    print(boto3.Session().region_name)

    is_running_on_sagemaker = "SM_INPUT_DATA_CONFIG" in os.environ
    if not is_running_on_sagemaker:
        est = PyTorch(
            entry_point=str(Path(__file__)),
            py_version="py3",
            framework_version='1.6',
            instance_count=1,
            instance_type="ml.m5.large",
            role=get_execution_role(),
            environment={"AWS_DEFAULT_REGION": "us-east-1"}
        )
        est.fit()
```

Reply:
Yes, you are right - there was a bug in my previous attempt that caused a wrong behavior. I'm opening a new PR with a refactored implementation.